team/bundles/org.eclipse.compare.core/src/org/eclipse/compare/rangedifferencer/RangeDifferencer.java

```diff
@@ -97,7 +97,64 @@ public static RangeDifference[] findDifferences(IProgressMonitor pm, IRangeCompa
 	 * @since org.eclipse.compare.core 3.5
 	 */
 	public static RangeDifference[] findDifferences(AbstractRangeDifferenceFactory factory, IProgressMonitor pm, IRangeComparator left, IRangeComparator right) {
-		return RangeComparatorLCS.findDifferences(factory, pm, left, right);
+		return balanceDifferences(RangeComparatorLCS.findDifferences(factory, pm, left, right), factory, left, right);
+	}
+
+	/**
+	 * Balance differences. The LCS computation tries to produce identical parts in
+	 * the comparison as large as possible. That is not necessarily what humans
+	 * expect.
+	 * <p>
+	 * Imagine the ranges ...ABCDABEF... where LCS has led to marking CDAB as
+	 * difference (because it tries to maximize the sequence before that change).
+	 * E.g. AB[CDAB]EF is the marked change. If the other side is an _empty_ change,
+	 * then humans typically prefer BCDA as change, e.g. A[BCDA]BEF. In fact,
+	 * anything from [ABCD]ABEF to AB[CDAB]EF is possible, and choosing the median
+	 * of all those possibilities should lead to better results if the input is for
+	 * instance structured text (with different indentation levels).
+	 *
+	 * @param diff    if not <code>null</code> used to report progress
+	 * @param factory factory to instantiate new {@link RangeDifference}s
+	 * @param left    the left range comparator
+	 * @param right   the right range comparator
+	 * @return an array of range differences, or an empty array if no differences
+	 *         were found
+	 */
+	private static RangeDifference[] balanceDifferences(RangeDifference[] diffs, AbstractRangeDifferenceFactory factory, IRangeComparator left,
+			IRangeComparator right) {
+		ArrayList<RangeDifference> result = new ArrayList<>();
+		int previousEnd = 0;
```

Comment:
`previousEnd` is never incremented, which I think, causes overlapping `RangeDifference`s:

We have a failing unit, which compares two sources and applies all differences:

```java
		MultiTextEdit result = new MultiTextEdit();

		TextRangeComparator left = new TextRangeComparator(fromText, compareMode);
		TextRangeComparator right = new TextRangeComparator(toText, compareMode);

		RangeDifference[] differences = RangeDifferencer.findDifferences(left, right);
		for (RangeDifference difference : differences) {
			TextRange leftRange = left.getRange(difference.leftStart(), difference.leftLength());
			TextRange rightRange = right.getRange(difference.rightStart(), difference.rightLength());

			ReplaceEdit replaceEdit = new ReplaceEdit(leftRange.offset, leftRange.length, rightRange.text);
			result.addChild(replaceEdit);
		}
		result.apply(document);

```
which is failing since this change:
```
org.eclipse.text.edits.MalformedTreeException: Overlapping text edits
	at org.eclipse.text.edits.TextEdit$InsertionComparator.compare(TextEdit.java:126)
	at org.eclipse.text.edits.TextEdit$InsertionComparator.compare(TextEdit.java:1)
	at java.base/java.util.Collections.indexedBinarySearch(Collections.java:336)
	at java.base/java.util.Collections.binarySearch(Collections.java:324)
	at org.eclipse.text.edits.TextEdit.computeInsertionIndex(TextEdit.java:799)
	at org.eclipse.text.edits.TextEdit.internalAdd(TextEdit.java:784)
	at org.eclipse.text.edits.TextEdit.addChild(TextEdit.java:340)
	...
```

Reply:
https://github.com/eclipse-platform/eclipse.platform/issues/458

I'll provide a fix proposal.

Reply:
https://github.com/eclipse-platform/eclipse.platform/pull/459